File: src/leetcode/python/two_pointers/ImplementStrStr_28.py
class ImplementStrStr_28(object):
    """
    The fastest way:　↓
    for i in range(len(haystack) - len(needle) + 1):
        if haystack[i:i + len(needle)] == needle:
            return i
    return -1
    """
    def strStr(self, haystack, needle):
        if len(haystack) < len(needle):
            return -1
        elif not len(haystack) or not len(needle):
            return 0

        for i in range(len(haystack) - len(needle) + 1):
            j = 0
            if haystack[i] == needle[j]:
                k = i
                while j < len(needle) and haystack[k] == needle[j]:
                    k += 1
                    j += 1
                if j == len(needle):
                    return i
        return -1

File: src/leetcode/java/easy/ImplementStrStr_28.java
package leetcode.java.easy;

public class ImplementStrStr_28 {
  public int strStr(String haystack, String needle) {
    char[] hay = haystack.toCharArray();
    char[] ned = needle.toCharArray();
    if (hay.length < ned.length) return -1;
    else if (hay.length == 0 || ned.length == 0) return 0;

    for (int i = 0, j = 0, k = 0; i < hay.length - ned.length + 1; i++) {
      if (hay[i] == ned[j]) {
        k = i;
        while (j < ned.length && hay[k] == ned[j]) {
          k++;
          j++;
        }
        if (j == ned.length) return i;
        j = 0;
      }
    }
    return -1;
  }
}
